common/src/interlocked_hl.c

```diff
@@ -100,20 +101,41 @@ IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_WaitForValue,
             {
                 /* Codes_SRS_INTERLOCKED_HL_01_003: [ If the value at address is equal to value, InterlockedHL_WaitForValue shall return INTERLOCKED_HL_OK. ]*/
                 result = INTERLOCKED_HL_OK;
+                continue_eval = false;
                 break;
             }
 
             /* Codes_SRS_INTERLOCKED_HL_01_008: [ If the value at address does not match, InterlockedHL_WaitForValue shall issue another call to wait_on_address. ]*/
 
             /* Codes_SRS_INTERLOCKED_HL_01_004: [ If the value at address is not equal to value, InterlockedHL_WaitForValue shall wait until the value at address changes in order to compare it again to value by using wait_on_address. ]*/
             /* Codes_SRS_INTERLOCKED_HL_01_005: [ When waiting for the value at address to change, the milliseconds argument value shall be used as timeout. ]*/
-            if (wait_on_address(address, current_value, milliseconds) != WAIT_ON_ADDRESS_OK)
+            WAIT_ON_ADDRESS_RESULT wait_result = wait_on_address(address, current_value, milliseconds);
+            switch (wait_result)
             {
-                /* Codes_SRS_INTERLOCKED_HL_01_006: [ If wait_on_address fails, InterlockedHL_WaitForValue shall fail and return INTERLOCKED_HL_ERROR. ]*/
-                result = INTERLOCKED_HL_ERROR;
-                break;
+                case WAIT_ON_ADDRESS_OK:
+                {
+                    result = INTERLOCKED_HL_OK;
+                    break;
```

Comment:
Yeah, so there's problems with the code here, and that's VERY unfortunate because these problems also point out to the tests being insufficient.

Here's what CAN happen:
line 82: InterlockedHL_WaitForValue is called with value=1
line 99: observes the value being "2". 
line 112: wait_on_address returns `WAIT_ON_ADDRESS_OK` but it is a spurious wake-up ([https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress)), so the value is still "2". Note: return is "WAIT_ON_ADDRESS_OK".
line 118: returns to the user "OK".

At no time during this execution did the value get to be "1". 

The fact that this is not captured by tests is concerning.

Reply:
I'm guessing the test writer did not think of surious wake-ups.  I'll add a test

Reply:
I will ... errr... reformulate my comment from above since it is factually inaccurate. 

As I've self-observed, `break` is overloaded (in another comment).

All i've said can be discarded but don't set "result" to anything because that's confusing :)

Reply:
After looking there is a test for this